[Quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73510
* #73486
* #73337
* __->__ #73035

Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API.
The support is implemented using optional variables and arguments.
We may wish to explore overloading raw_cudnn_convolution_forward_out
in the future as opposed to using optional variables, as that may be more performant.

Testing:
uncomment
```
    @unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

dzdang committed:
[quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn

Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API. The support is implemented using optional variables and arguments.

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34314535).

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -72,13 +72,14 @@ cudnn_frontend::ConvDesc_v8 getConvDescriptor(cudnnDataType_t dataType, IntArray
 // TODO: there is a table from input dtype to operator dtype, we can derive
 // the operator dtype based on input dtype
-cudnn_frontend::PointWiseDesc_v8 getPointWiseMulDescriptor(cudnnDataType_t dataType) {
+cudnn_frontend::PointWiseDesc_v8 getPointWiseDescriptor(cudnnDataType_t dataType, cudnnPointwiseMode_t pw_op_type) {
```
I feel it's still good to keep `getPointWiseMulDescriptor` so that we don't need to pass around `CUDNN_POINTWISE_MUL`, it can call this function.
We can create a `getPointWiseAddDescriptor` as well

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -358,17 +414,19 @@ class QConvInt8 final {
         output_scale,
-        output_zero_point
+        output_zero_point,
+        bias_multiplier
```
nit: probably add this before `requantize_multiplier`

dzdang committed:
Update on "[quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API. The support is implemented using optional variables and arguments.

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API. The support is implemented using optional variables and arguments.

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API.
The support is implemented using optional variables and arguments.
We may wish to explore overloading raw_cudnn_convolution_forward_out
in the future as opposed to using optional variables, as that may be more performant.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API.
The support is implemented using optional variables and arguments.
We may wish to explore overloading raw_cudnn_convolution_forward_out
in the future as opposed to using optional variables, as that may be more performant.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34314535).

dzdang committed:
Update on "[quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API.
The support is implemented using optional variables and arguments.
We may wish to explore overloading raw_cudnn_convolution_forward_out
in the future as opposed to using optional variables, as that may be more performant.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API.
The support is implemented using optional variables and arguments.
We may wish to explore overloading raw_cudnn_convolution_forward_out
in the future as opposed to using optional variables, as that may be more performant.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API.
The support is implemented using optional variables and arguments.
We may wish to explore overloading raw_cudnn_convolution_forward_out
in the future as opposed to using optional variables, as that may be more performant.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -208,20 +240,36 @@ void raw_cudnn_convolution_forward_out(
   key.input_alignment = getAlignment(input);
   key.output_alignment = getAlignment(conv_output);
   key.weight_alignment = getAlignment(weight);
+  if (bias.has_value()) key.bias_alignment = getAlignment(broadcasted_bias.value());
+  else key.bias_alignment = -1;
```
can we use:

```
if (cond) {

} else {

}
```

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -199,6 +215,22 @@ void raw_cudnn_convolution_forward_out(
+  c10::optional<at::Tensor> after_add;
+  c10::optional<at::Tensor> broadcasted_bias;
+  if (bias.has_value()) {
+    std::vector<int64_t> new_size(quantized_output.sizes().size() - 1, 1);
```
`quantized_output.sizes().size()` is the same as `quantized_output.ndim()` I think. 

also can we add some comment describing the broadcasting process as well, e.g. why we are using size() - 1 here etc.

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -199,6 +215,22 @@ void raw_cudnn_convolution_forward_out(
+    broadcasted_bias.value() = broadcasted_bias.value().contiguous(c10::MemoryFormat::ChannelsLast);
+    bias_multiplier_tensor = at::empty(quantized_output.sizes(), at::device(at::kCUDA).dtype(at::kFloat), at::MemoryFormat::ChannelsLast);
+    bias_multiplier_tensor.value().fill_(bias_multiplier);
+    after_scales_bias = at::empty(quantized_output.sizes(), at::device(at::kCUDA).dtype(at::kFloat), at::MemoryFormat::ChannelsLast);
```
btw, have we asked cudnn team about the support for inplace operations?

jerryzh168 replied:
saw this in the comments below

dzdang replied:
I haven't asked yet, and it doesn't look like it was in that doc you created either. I'll check with them the next time we email them

jerryzh168 replied:
sg, let's keep a list of questions that we want to ask in the doc as well, just to keep track of everything

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -233,25 +281,51 @@ void raw_cudnn_convolution_forward_out(
   // std::cout << "operator:" << requant_op.describe() << std::endl;
 
-  std::array<cudnn_frontend::Operation const *, 2> ops = {&conv_op, &requant_op};
+  std::vector<cudnn_frontend::Operation const *> ops{bias.has_value()};
```
where is conv_op? why do we have `bias.has_value()` here? is this supposed to be `&conv_op`?

dzdang replied:
yes, sorry, I forgot to ghstack the latest amended commit that fixed this. I'll get that up right away

dzdang commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -199,6 +215,22 @@ void raw_cudnn_convolution_forward_out(
   // TODO: compile empty & fill_ using full_like or full
   Tensor requantize_multiplier_tensor = at::empty(quantized_output.sizes(), at::device(at::kCUDA).dtype(at::kFloat), at::MemoryFormat::ChannelsLast);
   requantize_multiplier_tensor.fill_(requantize_multiplier);
+  c10::optional<at::Tensor> bias_multiplier_tensor;
```
yeah I think that would be a good idea. Should we put this on the pytorch doc site and refer them to there or put it in the comments? Might be easier to read math formulas on the site than in the comments

jerryzh168 replied:
since this is an implementation detail I feel it makes more sense to document this in the comments, we can either have inline comments around the operation graph, or for `raw_cudnn_convolution_forward` method (we can move it around later after the refactor to use packed params)

dzdang committed:
Update on "[quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API.
The support is implemented using optional variables and arguments.
We may wish to explore overloading raw_cudnn_convolution_forward_out
in the future as opposed to using optional variables, as that may be more performant.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API.
The support is implemented using optional variables and arguments.
We may wish to explore overloading raw_cudnn_convolution_forward_out
in the future as opposed to using optional variables, as that may be more performant.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API.
The support is implemented using optional variables and arguments.
We may wish to explore overloading raw_cudnn_convolution_forward_out
in the future as opposed to using optional variables, as that may be more performant.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -294,18 +373,35 @@ void raw_cudnn_convolution_forward_out(
+/*
+Numerics:
+Operation:
+out_fp32 = conv_fp32(act_fp32, w_fp32, …)
```
I think we also want to tie this closer to the code, we can write down the operations for each major op and their arguments, e.g. we have conv op, add bias op, requantize op etc.

jerryzh168 reviewed (approved):
looks great overall, had a few more inline comments

dzdang committed:
Update on "[Quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API.
The support is implemented using optional variables and arguments.
We may wish to explore overloading raw_cudnn_convolution_forward_out
in the future as opposed to using optional variables, as that may be more performant.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API.
The support is implemented using optional variables and arguments.
We may wish to explore overloading raw_cudnn_convolution_forward_out
in the future as opposed to using optional variables, as that may be more performant.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34314535).

dzdang committed:
Update on "[Quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API.
The support is implemented using optional variables and arguments.
We may wish to explore overloading raw_cudnn_convolution_forward_out
in the future as opposed to using optional variables, as that may be more performant.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API.
The support is implemented using optional variables and arguments.
We may wish to explore overloading raw_cudnn_convolution_forward_out
in the future as opposed to using optional variables, as that may be more performant.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34314535).

dzdang committed:
Update on "[Quant][core][gpu] Implemented support for bias in quantized conv operator in cudnn"


Summary: This PR extends https://github.com/pytorch/pytorch/pull/70622 which neglected the bias term. Here, the bias term is implemented using the support provided by cudnn's frontend API.
The support is implemented using optional variables and arguments.
We may wish to explore overloading raw_cudnn_convolution_forward_out
in the future as opposed to using optional variables, as that may be more performant.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34314535](https://our.internmc.facebook.com/intern/diff/D34314535)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34314535).
